pkg/telemetry/client.go

```diff
@@ -0,0 +1,274 @@
+	cmd.Flags().Visit(func(flag *pflag.Flag) {
+		// capture the boolean and empty flag values as is
+		if !hashRequired || flag.Value.String() == "" || flag.Value.Type() == "bool" {
+			flagMap[flag.Name] = flag.Value.String()
+		} else {
+			flagMap[flag.Name] = hashString(flag.Value.String())
+		}
+	})
+	if len(flagMap) != 0 {
+		jsonString, _ := json.Marshal(flagMap)
+		tc.currentOperationMetrics.Flags = string(jsonString)
+	}
+
+	return nil
+}
+
+func (tc *telemetryClient) updateMetricsForPlugin(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+
+	flagNames := TraverseFlagNames(args)
+	if len(flagNames) > 0 {
+		tc.currentOperationMetrics.Flags = flagNamesToJSONString(flagNames)
+	}
+
+	plugin := tc.pluginInfoFromCommand(cmd)
+	if plugin != nil {
+		tc.currentOperationMetrics.PluginName = plugin.Name
+		tc.currentOperationMetrics.PluginVersion = plugin.Version
+		tc.currentOperationMetrics.Target = string(plugin.Target)
+		tc.currentOperationMetrics.Endpoint = getEndpointSHA(plugin)
+	}
+
+	// TODO : Fix the command chain for plugins by using command chain cache constructed from generate-all-docs command
+	tc.currentOperationMetrics.CommandName = strings.Join(strings.Split(cmd.CommandPath(), " ")[1:], " ")
+
+	return nil
+}
+func (tc *telemetryClient) pluginInfoFromCommand(cmd *cobra.Command) *cli.PluginInfo {
+	var plugin *cli.PluginInfo
+
+	cmds := strings.Split(cmd.CommandPath(), " ")
+
+	// check if the second command in command path matches any target,
+	// if so, look for the 3rd command in the plugins map to find the plugin
+	cmdTarget := configtypes.StringToTarget(cmds[1])
+	if cmdTarget == configtypes.TargetK8s || cmdTarget == configtypes.TargetTMC {
+		if m, exists := tc.installedPluginsByTargetType[cmdTarget]; exists {
+			plugin = m[cmds[2]]
+		}
+	} else {
+		// currently global target plugin would contain the k8s target plugins since we are adding
+		// all the root level plugin commands(k8s plugins are root level at the moment) to
+		// "installedPluginsByTargetType" map with target type "global"
+		m := tc.installedPluginsByTargetType[configtypes.TargetGlobal]
+		if m != nil {
+			plugin = m[cmds[1]]
+		}
+	}
+	return plugin
+}
+
+func cliInstanceID() (string, error) {
+	cliID, err := configlib.GetCLIId()
+	if err != nil {
+		return "", err
+	}
+	return cliID, nil
+}
+
+func ensureMetricsSource() error {
+	telemetryOptions, _ := configlib.GetCLITelemetryOptions()
+	dbFile := filepath.Join(common.DefaultCLITelemetryDir, SQliteDBFileName)
+	if telemetryOptions != nil && telemetryOptions.Source == dbFile {
+		return nil
+	}
+
+	err := configlib.SetCLITelemetryOptions(&configtypes.TelemetryOptions{Source: dbFile})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// isHashRequiredForCmdFlags determines if hashing is required for a core command read from command path
```

Comment:
Nit: I suggest adding a comment about the limitation of this check, which is that for each command we are either hashing all values or none at all. A possible enhancement would be to return a list of flags whose values need to be hashed.

Reply:
updated.